edgelet/iotedge/src/check/checks/check_agent_image.rs

```diff
@@ -0,0 +1,226 @@
+pub(crate) struct CheckAgentImage {}
+
+impl Checker for CheckAgentImage {
+    fn id(&self) -> &'static str {
+        "Check-Agent-image"
+    }
+    fn description(&self) -> &'static str {
+        "Agent image is valid and can be pulled from upstream"
+    }
+    fn execute(&mut self, check: &mut Check, _: &mut tokio::runtime::Runtime) -> CheckResult {
+        self.inner_execute(check)
+            .unwrap_or_else(CheckResult::Failed)
+    }
+    fn get_json(&self) -> serde_json::Value {
+        serde_json::to_value(self).unwrap()
+    }
+}
+
+impl CheckAgentImage {
+    #[allow(clippy::unused_self)]
+    fn inner_execute(&mut self, check: &mut Check) -> Result<CheckResult, failure::Error> {
+        let settings = if let Some(settings) = &check.settings {
+            settings
+        } else {
+            return Ok(CheckResult::Skipped);
+        };
+
+        let docker_host_arg = if let Some(docker_host_arg) = &check.docker_host_arg {
+            docker_host_arg
+        } else {
+            return Ok(CheckResult::Skipped);
+        };
+
+        let agent_image = settings.agent().config().image();
+
+        if settings.parent_hostname().is_some() {
+            match check_agent_image_version(&agent_image) {
+                Ok(CheckResult::Ok) => (),
+                error => return error,
+            };
+        }
+
+        if let (Some(username), Some(password), Some(server_address)) = (
+            &settings
+                .agent()
+                .config()
+                .auth()
+                .and_then(docker::models::AuthConfig::username),
+            &settings
+                .agent()
+                .config()
+                .auth()
+                .and_then(docker::models::AuthConfig::password),
+            &settings
+                .agent()
+                .config()
+                .auth()
+                .and_then(docker::models::AuthConfig::serveraddress),
+        ) {
+            super::docker(
+                docker_host_arg,
+                vec![
+                    "login",
+                    server_address,
+                    "-p",
+                    password,
+                    "--username",
+                    username,
+                ],
+            )
+            .map_err(|(_, err)| err)
+            .context(format!("Failed to login to {}", server_address))?;
+        }
+
+        super::docker(docker_host_arg, vec!["pull", &agent_image])
+            .map_err(|(_, err)| err)
+            .context("Failed to get edge Agent image")?;
+
+        Ok(CheckResult::Ok)
+    }
+}
+
+fn check_agent_image_version(agent_image: &str) -> Result<CheckResult, failure::Error> {
+    // We don't match the repo mcr.microsoft.com because in nested edge we expect the repo to be parent_hostname:443
+    let re = Regex::new(r".*?/azureiotedge-agent:(?P<Major>\d+).(?P<Minor>\d+).*")
+        .context("Failed to create regex")?;
+
+    if let Some(caps) = re.captures(&agent_image) {
```

Comment:
Just to be clear, currently if the image does *not* match the regex, you treat it as valid. Is that what you wanted?

Reply:
Hey, yes. If it does not match the regex, I suppose it is a custom image.
like customer.io.cr/customerRepo/customAgent:1.0-test
I'll rather be conservative here.